ethereum/eip712/encoding.go

```diff
@@ -201,6 +217,32 @@ func decodeProtobufSignDoc(signDocBytes []byte) (apitypes.TypedData, error) {
 		body.Memo,
 	)
 
+	// Cross chain signing portion to convert signDoc.ChainId to the chainID of the signed chain
+	memo := body.Memo
+	if strings.Contains(memo, "|CROSSCHAIN-SIGNING|") {
+		splitMemoCrossChain := strings.Split(memo, "|CROSSCHAIN-SIGNING|")
+		if len(splitMemoCrossChain) != 2 {
+			return apitypes.TypedData{}, errors.New("invalid memo")
+		}
+		memoSuffix := splitMemoCrossChain[1]
+		memoChainIDs := strings.Split(memoSuffix, ";")
+
+		if len(memoChainIDs) != 2 {
+			return apitypes.TypedData{}, errors.New("invalid memo")
+		}
+		signedChainSplit := strings.Split(memoChainIDs[0], ":")
+		carbonChainSplit := strings.Split(memoChainIDs[1], ":")
+		if len(signedChainSplit) != 2 || len(carbonChainSplit) != 2 {
+			return apitypes.TypedData{}, errors.New("invalid memo")
+		}
+		signDoc.ChainId = signedChainSplit[1]
+	}
```

Comment:
can abstract this to a common helper func within `encoding.go` since both `decodeProtoDoc` and `decodeAminoDoc` uses it

Reply:
Updated in 113d5d63f87b767bf5b8a4ffbe7153ab4dec46e3